client/identities.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+)
+
+// Identity holds the configuration of a single identity.
+type Identity struct {
+	Access IdentityAccess `json:"access" yaml:"access"`
+
+	// One or more of the following type-specific configuration fields must be
+	// non-nil (currently the only type is "local").
+	Local *LocalIdentity `json:"local,omitempty" yaml:"local,omitempty"`
+}
+
+// IdentityAccess defines the access level for an identity.
+type IdentityAccess string
+
+const (
+	AdminAccess     IdentityAccess = "admin"
+	ReadAccess      IdentityAccess = "read"
+	UntrustedAccess IdentityAccess = "untrusted"
+)
+
+// LocalIdentity holds identity configuration specific to the "local" type
+// (for ucrednet/UID authentication).
+type LocalIdentity struct {
+	// This is a pointer so we can distinguish between not set and 0 (a valid
+	// user-id meaning root).
+	UserID *uint32 `json:"user-id" yaml:"user-id"`
+}
+
+// For future extension.
+type IdentitiesOptions struct{}
+
+type identitiesPayload struct {
+	Action     string               `json:"action"`
+	Identities map[string]*Identity `json:"identities"`
+}
+
+// Identities returns a map of all identities in the system.
+func (client *Client) Identities(opts *IdentitiesOptions) (map[string]*Identity, error) {
+	resp, err := client.Requester().Do(context.Background(), &RequestOptions{
+		Type:   SyncRequest,
+		Method: "GET",
+		Path:   "/v1/identities",
+	})
+	if err != nil {
+		return nil, err
+	}
+	var identities map[string]*Identity
+	err = resp.DecodeResult(&identities)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We should probably be kind to ourselves and check the server did not return a nil entry in the map.

Reply:
Fair enough -- added this and a test.